core/store/src/contract.rs

```diff
@@ -24,61 +24,159 @@ pub struct ContractStorage {
     /// The State this field should be removed, and the `Storage::store` function should be
     /// adjusted to write out the contract into the relevant part of the database immediately
     /// (without going through transactional storage operations and such).
-    uncommitted_deploys: Arc<Mutex<BTreeMap<CryptoHash, ContractCode>>>,
+    uncommitted_deploys: BTreeMap<CodeHash, ContractCode>,
+
+    /// Deployments moved from `uncommitted_deploys` by calling `commit`.
+    /// These are returned from `finalize`.
+    committed_deploys: BTreeMap<CodeHash, ContractCode>,
+
+    /// List of code-hashes for the contracts called.
+    ///
+    /// We do not distinguish between committed and uncommitted calls, because we need
+    /// to record all calls to validate both successful and failing function calls.
+    contract_calls: BTreeSet<CodeHash>,
+}
+
+impl ContractsTracker {
+    fn get(&self, code_hash: CodeHash) -> Option<ContractCode> {
+        self.uncommitted_deploys
+            .get(&code_hash)
+            .or_else(|| self.committed_deploys.get(&code_hash))
+            .and_then(|contract| {
+                Some(ContractCode::new(contract.code().to_vec(), Some(code_hash.into())))
+            })
+    }
 
-    /// List of code-hashes for the contracts retrieved from the storage.
-    /// This does not include the contracts deployed and then read.
-    // TODO(#11099): For collecting deployed code, consolidate this with uncommitted_deploys.
-    storage_reads: Arc<Mutex<Option<HashSet<CodeHash>>>>,
+    fn call(&mut self, code_hash: CodeHash) {
+        self.contract_calls.insert(code_hash);
+    }
+
+    fn deploy(&mut self, code: ContractCode) {
+        self.uncommitted_deploys.insert((*code.hash()).into(), code);
+    }
+
+    /// Rollsback the uncommitted deployments.
+    fn rollback_deploys(&mut self) {
+        self.uncommitted_deploys.clear();
+    }
+
+    /// Commits the uncommitted deployments by moving them to the set of committed deployments and clearing the uncommitted list.
+    fn commit_deploys(&mut self) {
+        let deploys = std::mem::take(&mut self.uncommitted_deploys);
+        for (code_hash, contract) in deploys.into_iter() {
+            self.committed_deploys.insert(code_hash, contract);
+        }
+    }
+
+    /// Finalizes this tracker and returns the calls and committed deployments.
+    fn finalize(mut self) -> ContractStorageResult {
+        ContractStorageResult {
+            contract_calls: std::mem::take(&mut self.contract_calls),
+            contract_deploys: self.committed_deploys.into_keys().collect(),
+        }
+    }
 }
 
-/// Result of finalizing the contract storage.
+/// Result of finalizing the contract storage, containing the contract calls and committed deployments.
 pub struct ContractStorageResult {
-    pub contract_accesses: Vec<CodeHash>,
+    /// List of code-hashes for the contract calls while applying the chunk.
+    pub contract_calls: BTreeSet<CodeHash>,
```

Comment:
why do want to have `BTreeSet`s here instead of  just `Vec`?

Reply:
see my other comment about using set for uniqueness. Note that I could also use a HashMap, but since I was collecting them in BTreeSet in the ContractStorage, I decided to continue passing it around.